compare and display package versions

emctackett opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

@MikeMcQuaid this is a follow-up from our interaction (including @baffouradu as well) in the MLH discord server, drafting a PR here as suggested.

Initially, this code was created in response to this issue: https://github.com/Homebrew/brew/issues/5725 .. it has since been modified according to your feedback.

For now I placed the code from our repository into the `scripts` dir; you will see that the `README.md` is included.  As-is, when `ruby printPackageUpdates.rb` is run, the code queries homebrew, repology, and livecheck to figure out which packages are outdated and what the latest version available is.  The code also checks whether there is an existing PR for the package.  These results are output to the user.

Example output:
```
Package: openclonk
Brew current: 7.0
Repology latest: 8.1
Livecheck latest: 8.1
Has Open PR?: true

Package: openjdk
Brew current: 13.0.2+8
Repology latest: 15.0.0.0~14
Livecheck latest: Not found.
Has Open PR?: false

Package: opentsdb
Brew current: 2.3.1
Repology latest: 2.4.0
Livecheck latest: 2.4.0
Has Open PR?: true
```

We welcome your feedback, direction, or questions regarding next steps. 

Thanks!

MikeMcQuaid commented:
> We welcome your feedback, direction, or questions regarding next steps.

Thanks, nice work so far! I'm thinking that there be a `dev-cmd/bump.rb` or something that outputs similar to the above output (which we can iterate on). 

Thoughts on output:

> Package: openclonk

`Formula: openclonk` or just use `ohai` method which will do `==> openclonk`

> Brew current: 7.0

Current formula version: 7.0 (and get this from `Formula["openclonk"].version` rather than HTTP call)

> Repology latest: 8.1

Latest repology version: 8.1

> Livecheck latest: 8.1

Latest livecheck version: 8.1

> Has Open PR?: true

Open pull request: https://url/to/pullrequest

Could also consider outputting the `brew bump-formula-pr` line for copy-pasting if there's no open pull request? Would also probably want to see closed pull requests with an exact version match (e.g. https://github.com/Homebrew/brew/pull/7745)

---

@Homebrew/maintainers What do you thoughts on my thoughts above and whether this should be one/multiple commands and what output would be most useful to you? The thinking is this would be something that initially used to inform contributors to be able to quickly know what new pull requests to create and eventually could be used with some automation. CC @samford in particular due to livecheck work and the livecheck GSoC.

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,47 @@
+  def call_api(url)
+    puts "- Calling API #{url}"
+    uri = URI(url)
+    response = Net::HTTP.get(uri)
```
Use the `curl` method rather than `Net::HTTP` to access network resources.

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,47 @@
+  end
+
+  def parse_repology_api
+    puts "\n-------- Query outdated packages from Repology --------"
```
```suggestion
    ohai "Querying outdated packages from Repology"
```

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,47 @@
+
+module RepologyParser
+  def call_api(url)
+    puts "- Calling API #{url}"
```
```suggestion
    ohai "Calling API #{url}" if Homebrew.args.verbose?
```

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,47 @@
+    uri = URI(url)
+    response = Net::HTTP.get(uri)
+
+    puts "- Parsing response"
```
```suggestion
    ohai "Parsing response" if Homebrew.args.verbose?
```

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,47 @@
+  def parse_repology_api
+    puts "\n-------- Query outdated packages from Repology --------"
+    page_no = 1
+    puts "\n- Paginating repology api page: #{page_no}"
```
Use `ohai` (and everywhere else like this)

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,47 @@
+    puts "\n- Paginating repology api page: #{page_no}"
+
+    outdated_packages = query_repology_api("")
+    last_pacakge_index = outdated_packages.size - 1
```
```suggestion
    last_package_index = outdated_packages.size - 1
```

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,47 @@
+    page_no = 1
+    puts "\n- Paginating repology api page: #{page_no}"
+
+    outdated_packages = query_repology_api("")
```
Why `""`?

BaffourAdu replied:
@MikeMcQuaid this is because of the way Repology paginates it's API

MikeMcQuaid replied:
Sorry: I mean given it's the default argument: why pass it here?

emctackett replied:
I'm removing this.  Thanks for pointing it out, Mike!

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,47 @@
+    last_pacakge_index = outdated_packages.size - 1
+    response_size = outdated_packages.size
+
+    while response_size > 1
```
Is there anything in the API response that provides a number of pages?

BaffourAdu replied:
@MikeMcQuaid there's nothing in the API to paginate the response. The API limits the results to 200 packages per response and to paginate through the API, you have to use the last package in every response to get the next set of response till you have only one package in the response which indicates the last page.

MikeMcQuaid replied:
Would be good to have a maximum limit so this loop doesn't get "stuck" forever.

BaffourAdu replied:
Noted! @emctackett and I have agreed to limit it to a maximum of 15 paginations.

MikeMcQuaid commented on scripts/README.md:
```diff
@@ -0,0 +1,47 @@
+# 0.0.1-API-Parser
```
Good to have this in your repo but don't need this in this PR.

MikeMcQuaid commented on scripts/bumpFormulae.rb:
```diff
@@ -0,0 +1,20 @@
+require_relative 'helpers/parsed_file'
```
This file should also be a `dev-cmd`

MikeMcQuaid commented on scripts/helpers/api_parser.rb:
```diff
@@ -0,0 +1,133 @@
+require 'net/http'
```
Double-quotes. Use `brew style --fix` to check and fix these all.

MikeMcQuaid commented on scripts/helpers/api_parser.rb:
```diff
@@ -0,0 +1,133 @@
+require_relative 'brew_commands'
+require_relative 'homebrew_formula'
+
+class ApiParser
```
Looks like this duplicates `utils/repology` above.

sjackman commented:
On the output format, in addition to the plain text format, one structured format such as either JSON or TSV would be useful. Probably JSON, as that's what other Homebrew tools output.

I don't know if it was intentional or a happy accident, but Miller (`mlr`) can convert the proposed plain text format to JSON or TSV.

```bash
$ brew install miller
$ mlr --ixtab --ips ': ' --ojson cat <<EOS
Package: openclonk
Brew current: 7.0
Repology latest: 8.1
Livecheck latest: 8.1
Has Open PR?: true

Package: openjdk
Brew current: 13.0.2+8
Repology latest: 15.0.0.0~14
Livecheck latest: Not found.
Has Open PR?: false
EOS
{ "Package": "openclonk", "Brew current": 7.0, "Repology latest": 8.1, "Livecheck latest": 8.1, "Has Open PR?": true }
{ "Package": "openjdk", "Brew current": "13.0.2+8", "Repology latest": "15.0.0.0~14", "Livecheck latest": "Not found.", "Has Open PR?": false }
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,24 @@
+  def bump
+    bump_args.parse
+    # puts "command run"
+    outdated_repology_pacakges = RepologyParser.parse_api_response()
```
```suggestion
    outdated_repology_packages = RepologyParser.parse_api_response
```

MikeMcQuaid commented on Library/Homebrew/utils/livecheck.rb:
```diff
@@ -0,0 +1,28 @@
+
+module Livecheck
+  def livecheck_formula_response(formula_name)
+    ohai "- livecheck formula : #{formula_name}"
```
```suggestion
    ohai "checking livecheck formula : #{formula_name}" 
```

or something? Describe what you're doing

MikeMcQuaid commented on Library/Homebrew/utils/livecheck.rb:
```diff
@@ -0,0 +1,28 @@
+module Livecheck
+  def livecheck_formula_response(formula_name)
+    ohai "- livecheck formula : #{formula_name}"
+    command_args = [
```
inline these

MikeMcQuaid commented on Library/Homebrew/utils/livecheck.rb:
```diff
@@ -0,0 +1,28 @@
+      "--quiet",
+    ]
+
+    response = Open3.capture2e(*command_args)
```
What are you trying to capture here? Use `Utils.popen_read`, probably.

MikeMcQuaid commented on Library/Homebrew/utils/livecheck.rb:
```diff
@@ -0,0 +1,28 @@
+  end
+
+  def parse_livecheck_response(response)
+    output = response.first.delete(" ").split(/:|==>|\n/)
```
Why `.first`?

emctackett replied:
This is because when we were using `Open3.capture2e`, the response was a 2 element array, with the second element being the process status.  

Since we are using `Utils.popen_read` now, I'l remove it.

MikeMcQuaid commented on Library/Homebrew/utils/livecheck.rb:
```diff
@@ -0,0 +1,28 @@
+    # eg: ["burp", "2.2.18", "2.2.18"]
+    package_name, brew_version, latest_version = output
+
+    { "name" => package_name, "current_brew_version" => brew_version,
+      "livecheck_latest_version" => latest_version }
```
```suggestion
    { 
      name: package_name, 
      formula_version: brew_version,
      livecheck_version: latest_version,
    }
```

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,75 @@
+  def parse_api_response()
+    ohai "Querying outdated packages from Repology"
+    page_no = 1
+    ohai "Paginating repology api page: #{page_no}" if Homebrew.args.verbose?
```
```suggestion
    ohai "Paginating Repology api page: #{page_no}" if Homebrew.args.verbose?
```

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,75 @@
+    page_no = 1
+    ohai "Paginating repology api page: #{page_no}" if Homebrew.args.verbose?
+
+    outdated_packages = query_api()
```
```suggestion
    outdated_packages = query_api
```

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,75 @@
+    ohai "Paginating repology api page: #{page_no}" if Homebrew.args.verbose?
+
+    outdated_packages = query_api()
+    last_pacakge_index = outdated_packages.size - 1
```
```suggestion
    last_package_index = outdated_packages.size - 1
```

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,75 @@
+    outdated_packages = query_api()
+    last_pacakge_index = outdated_packages.size - 1
+    response_size = outdated_packages.size
+    page_limit = 15
```
Make this a constant.

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,75 @@
+
+    while response_size > 1 && page_no <= page_limit
+      page_no += 1
+      ohai "Paginating repology api page: #{page_no}" if Homebrew.args.verbose?
```
```suggestion
      ohai "Paginating Repology api page: #{page_no}" if Homebrew.args.verbose?
```

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,75 @@
+      ohai "Paginating repology api page: #{page_no}" if Homebrew.args.verbose?
+
+      last_package_in_response = outdated_packages.keys[last_pacakge_index]
+      response = query_api("#{last_package_in_response}/")
```
Feels weird to have to pass the `/` here. That should be done by the called function, I think.

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,75 @@
+
+      response_size = response.size
+      outdated_packages.merge!(response)
+      last_pacakge_index = outdated_packages.size - 1
```
```suggestion
      last_package_index = outdated_packages.size - 1
```

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,75 @@
+  end
+
+  def validate__packages(outdated_repology_packages)
+    ohai "Verifying outdated repology packages as Homebrew Formulae"
```
```suggestion
    ohai "Verifying outdated Repology packages as Homebrew Formulae"
```

MikeMcQuaid commented on Library/Homebrew/utils/versions.rb:
```diff
@@ -0,0 +1,36 @@
+  end
+
+  def bump_formula_pr(formula_name, url)
+    command_args = [
```
Inline these arguments

MikeMcQuaid commented on Library/Homebrew/utils/versions.rb:
```diff
@@ -0,0 +1,36 @@
+      "--url=#{url}",
+    ]
+
+    response = Open3.capture2e(*command_args)
```
Use `Utils.popen_read`

MikeMcQuaid commented on Library/Homebrew/utils/versions.rb:
```diff
@@ -0,0 +1,36 @@
+    ohai "- Checking for open PRs for formula : #{formula_name}"
+
+    response = bump_formula_pr(formula_name, download_url)
+    !response.include? "Error: These open pull requests may be duplicates"
```
Would be nice to do refactor this check into a separate function and call it rather than shelling out to `bump-formula-pr` to get this.

MikeMcQuaid commented on scripts/bumpFormulae.rb:
```diff
@@ -0,0 +1,20 @@
+require_relative "helpers/parsed_file"
```
I think these `script/` files can all be removed?

emctackett commented:
Thanks for the excellent feedback!  Will work on implementing this all.

emctackett commented on Library/Homebrew/utils/livecheck.rb:
```diff
@@ -0,0 +1,23 @@
+  def livecheck_formula_response(name)
+    ohai "Checking livecheck formula : #{name}"
+
+    response = Utils.popen_read("brew", "livecheck", name, "--quiet").chomp
```
Hi @MikeMcQuaid 

Today in Discord we discussed with you the best way to query livecheck from within our code, and also updated the code to use `Utils.popen_read` as suggested and seen above.  Our understanding from your comments was that we would need to use `brew livecheck` in order to livecheck a formula.  We have run into a bit of a blocker I'm hoping you can advise on.

Line 7 above, `Utils.popen_read("brew", "livecheck", name, "--quiet").chomp` works just fine if we run this `livecheck.rb` file.  Example output if we display the response:
```
ruby Homebrew/utils/livecheck.rb
# . . .
"openclonk : 7.0 ==> 8.1"
# . . 
```
However when this code is run as a consequence the `brew bump` command @BaffourAdu is working on implementing (in `dev-cmd/bump.rb`), an empty string is returned.  

#### What we've tried:
- We tried replacing "livecheck" with another brew command-- `Utils.popen_read("brew", "release-notes")`, which also returns an empty string.
- We tried replacing our arguments with something unrelated to brew-- `Utils.popen_read("git", "status")` and `Utils.popen_read('ls')`, and we *do* see the expected output.
- Paired with Matthew Bunday to see if he had any ideas on how to address this.
- Looked through the existing `Homebrew/brew` codebase and couldn't find any examples to emulate in which a `brew` command is being executed like this.

Given that we see the expected output when the arguments passed to `Utils.popen_read` are not a `brew` command, we suspect there is some difficulty with trying to livecheck the formulae via `brew livecheck` as a part of the execution of our `brew bump` command.

I hope this was clear, & we appreciate any guidance you can provide on this issue.  Thank you!

MikeMcQuaid replied:
Repeating from Discord just so it's here: try `Utils.popen_read(HOMEBREW_BREW_FILE, "livecheck", name, "--quiet")`, that should work.

jonchang commented on Library/Homebrew/utils/versions.rb:
```diff
@@ -0,0 +1,56 @@
+    }
+  end
+
+  def fetch_pull_requests(query, tap_full_name, state: nil)
+    GitHub.issues_for_formula(query, tap_full_name: tap_full_name, state: state).select do |pr|
+      pr["html_url"].include?("/pull/") &&
+        /(^|\s)#{Regexp.quote(query)}(:|\s|$)/i =~ pr["title"]
+    end
+  rescue GitHub::RateLimitExceededError => e
+    opoo e.message
+    []
+  end
```
This function should probably live in `utils/github.rb` and `dev-cmd/bump-formula-pr.rb` should be refactored to use it there to reduce code duplication.

MikeMcQuaid replied:
Agreed.

jonchang commented on Library/Homebrew/utils/versions.rb:
```diff
@@ -0,0 +1,56 @@
+    []
+  end
+
+  def check_for_duplicate_pull_requests(formula, version)
+    formula = Formula[formula]
+    tap_full_name = formula.tap&.full_name
+
+    # check for open requests
+    pull_requests = fetch_pull_requests(formula.name, tap_full_name, state: "open")
+
+    # if we haven't already found open requests, try for an exact match across all requests
+    pull_requests = fetch_pull_requests("#{formula.name} #{version}", tap_full_name) if pull_requests.blank?
+    return if pull_requests.blank?
+
+    pull_requests.map { |pr| { title: pr["title"], url: pr["html_url"] } }
+  end
```
Can we write this function such that `bump-formula-pr` and the new `bump` command can both use this?

MikeMcQuaid replied:
Agreed.

jonchang commented on Library/Homebrew/utils/versions.rb:
```diff
@@ -0,0 +1,56 @@
+  end
+
+  def livecheck_formula(formula)
+    ohai "Checking livecheck formula : #{formula}" if Homebrew.args.verbose?
```
```suggestion
    ohai "Checking livecheck formula: #{formula}" if Homebrew.args.verbose?
```

MikeMcQuaid replied:
Agreed.

jonchang commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,84 @@
+    last_package_in_response += "/" unless last_package_in_response.empty?
+
+    url = "https://repology.org/api/v1/projects/#{last_package_in_response}?inrepo=homebrew&outdated=1"
+    ohai "Calling API #{url}" if Homebrew.args.verbose?
```
```suggestion
    puts "Calling API #{url}" if Homebrew.args.verbose?
```

Though I'm not sure if passing `--verbose` will emit the `curl` URL anyway, so it might be good to check to see if this is even needed.

jonchang commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,84 @@
+    page_no = 1
+
+    while response_size > 1 && page_no <= MAX_PAGINATION
+      ohai "Paginating Repology api page: #{page_no}" if Homebrew.args.verbose?
```
```suggestion
      odebug "Paginating Repology API page: #{page_no}"
```

I feel this is more of a "debug" message than a "verbose" message.

jonchang commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,84 @@
+
+require "formula_info"
+
+module RepologyParser
```
Should this just be called `Repology` to match the file name?

jonchang commented on Library/Homebrew/utils/versions.rb:
```diff
@@ -0,0 +1,56 @@
+  module_function
+
+  def current_formula_version(formula_name)
+    Formula[formula_name].version.to_s.to_f
```
This will truncate the formula version if it has any non-numeric components, or a three-part semver version. Suggest you work directly with the `Version` object when possible.

```console
% brew irb
==> Interactive Homebrew Shell
Example commands available with: brew irb --examples

irb(main):001:0> "x264".f.version.to_s
=> "r2999"
irb(main):002:0> "x264".f.version.to_s.to_f
=> 0.0
irb(main):003:0> "curl".f.version.to_s
=> "7.71.1"
irb(main):004:0> "curl".f.version.to_s.to_f
=> 7.71
```

MikeMcQuaid replied:
Agreed.

jonchang commented on Library/Homebrew/utils/versions.rb:
```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+module Versions
```
I think most of the stuff in this file could just live directly in `dev-cmd/bump.rb`. We also already have a `version.rb` and I worry that this could get confusing.

MikeMcQuaid replied:
Agreed.

MikeMcQuaid commented on Library/Homebrew/.rubocop.yml:
```diff
@@ -2,12 +2,12 @@ inherit_from: ../.rubocop_rspec.yml
 
 AllCops:
   Include:
-    - '**/*.rb'
```
Leave these as-is, thanks!

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump-formula-pr.rb:
```diff
@@ -314,7 +314,7 @@ def bump_formula_pr
     new_formula_version = formula_version(formula, requested_spec, new_contents)
 
-    check_for_duplicate_pull_requests(formula, tap_full_name, new_formula_version.to_s)
+    GitHub.check_for_duplicate_pull_requests(formula, tap_full_name, new_formula_version.to_s, args)
```
```suggestion
    GitHub.check_for_duplicate_pull_requests(formula, tap_full_name, new_formula_version.to_s)
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,115 @@
+      repology_homebrew_repo = repositories.find do |repo|
+        repo["repo"] == "homebrew"
+      end
+
+      next if repology_homebrew_repo.empty?
```
```suggestion
      next if repology_homebrew_repo.blank?
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,115 @@
+      latest_version = nil
+
+      # identify latest version amongst repology repos
+      repositories.each do |repo|
```
Could this be `find`? Would it be set more than once?

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,115 @@
+
+  def format_package(package_name, latest_version)
+    formula = get_formula_details(package_name)
+
+    return if formula.nil?
```
```suggestion
    return if formula.blank?
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,115 @@
+  def parse_livecheck_response(response)
+    output = response.delete(" ").split(/:|==>/)
+
+    # eg: ["openclonk", "7.0", "8.1"]
```
```suggestion
    # e.g. ["openclonk", "7.0", "8.1"]
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,115 @@
+
+    outdated_packages.each do |formula, package_details|
+      ohai formula
+      ohai "Current formula version: #{package_details[:current_formula_version]}"
+      ohai "Latest repology version: #{package_details[:repology_latest_version]}"
+      ohai "Latest livecheck version: #{package_details[:livecheck_latest_version]}"
+      ohai "Open pull requests: #{package_details[:open_pull_requests]}"
```
```suggestion
      puts "Current formula version: #{package_details[:current_formula_version]}"
      puts "Latest repology version: #{package_details[:repology_latest_version]}"
      puts "Latest livecheck version: #{package_details[:livecheck_latest_version]}"
      puts "Open pull requests: #{package_details[:open_pull_requests]}"
```

MikeMcQuaid commented on Library/Homebrew/utils/github.rb:
```diff
@@ -337,6 +337,43 @@ def print_pull_requests_matching(query)
+    []
+  end
+
+  def check_for_duplicate_pull_requests(formula, tap_full_name, version, args, fetch_pr = false)
```
```suggestion
  def check_for_duplicate_pull_requests(formula, tap_full_name, version, fetch_pr = false)
```

And use `Homebrew.args` throughout. Alternatively, pass through `force:` and `quiet:` arguments.

MikeMcQuaid commented on Library/Homebrew/utils/github.rb:
```diff
@@ -337,6 +337,43 @@ def print_pull_requests_matching(query)
+    elsif !args.force? && args.quiet?
+      odie error_message
+    elsif !args.force?
+      odie <<~EOS
```
Are you sure you want this to `odie`and exit here?

BaffourAdu replied:
@MikeMcQuaid yes please, that functionality is for `bump-formula-pr` command which kills the script and outputs a duplicate error message.

MikeMcQuaid replied:
This doesn't seem desirable for `brew bump`, though? I'd suggest that this method removes the `fetch_pr` arguments, always returns `pull_requests.map ...` above and you move the `duplicates_message` and below logic back into `bump-formula-pr`

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,43 @@
+  MAX_PAGINATION = 15
+
+  def query_api(last_package_in_response = "")
+    last_package_in_response += "/" unless last_package_in_response.empty?
```
```suggestion
    last_package_in_response += "/" unless last_package_in_response.blank?
```

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,43 @@
+    page_no = 1
+
+    while response_size > 1 && page_no <= MAX_PAGINATION
+      odebug "Paginating Repology api page: #{page_no}"
```
```suggestion
      odebug "Paginating Repology API page: #{page_no}"
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,110 @@
+  end
+
+  def current_formula_version(formula)
+    formula.version.to_s
```
When does this fail? It never should.

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,43 @@
+  MAX_PAGINATION = 15
+
+  def query_api(last_package_in_response = "")
+    last_package_in_response += "/" unless last_package_in_response.blank?
```
```suggestion
    last_package_in_response += "/" if last_package_in_response.present?
```

emctackett commented:
@zencephalon -- tagging our MLH mentor Matthew here for some feedback on our work as well.

zencephalon commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,108 @@
+
+      latest_version = repositories.find { |repo| repo["status"] == "newest" }["version"]
+
+      packages[repology_homebrew_repo["srcname"]] = format_package(repology_homebrew_repo["srcname"], latest_version)
```
maybe we could `srcname = repology_homebrew_repo["srcname"]` to avoid accessing the hash twice.

zencephalon commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,108 @@
+  end
+
+  def parse_livecheck_response(response)
+    output = response.delete(" ").split(/:|==>/)
```
Defer to maintainers here. I like to include a comment with an example of `response` to help illustrate what the `delete` and `split` do.

zencephalon commented on Library/Homebrew/diagnostic.rb:
```diff
@@ -670,6 +670,11 @@ def check_git_status
 
+          # these will result in uncommitted gems.
+          if path == HOMEBREW_REPOSITORY
+            next if ENV["HOMEBREW_SORBET"] || ENV["HOMEBREW_PATCHELF_RB"]
```
Maybe `next if path == HOMEBREW_REPOSITORY && (ENV["HOMEBREW_SORBET"] || ENV["HOMEBREW_PATCHELF_RB"])`

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump-formula-pr.rb:
```diff
@@ -314,6 +314,8 @@ def bump_formula_pr
 
     new_formula_version = formula_version(formula, requested_spec, new_contents)
 
+    check_duplicate_pull_requests(formula, tap_full_name, new_formula_version.to_s)
```
How about extracting the `Library/Homebrew/dev-cmd/bump-formula-pr.rb` and `Library/Homebrew/utils/github.rb` changes into a separate PR? They could be merged immediately.

BaffourAdu replied:
Sure!

BaffourAdu replied:
@MikeMcQuaid I just realized this has already been done here => https://github.com/Homebrew/brew/commit/f684a59fa5f8dfe6c1498bdd35d08d1e1ab60e92

MikeMcQuaid replied:
Nice. Worth rebasing or merging to include those changes into here.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,110 @@
+  end
+
+  def display(outdated_packages)
+    ohai "Outdated formulae\n"
```
```suggestion
    ohai "Outdated formulae"
    puts
```

MikeMcQuaid commented on Library/Homebrew/diagnostic.rb:
```diff
@@ -671,9 +671,7 @@ def check_git_status
-          if path == HOMEBREW_REPOSITORY
-            next if ENV["HOMEBREW_SORBET"] || ENV["HOMEBREW_PATCHELF_RB"]
-          end
+          next if path == HOMEBREW_REPOSITORY && (ENV["HOMEBREW_SORBET"] || ENV["HOMEBREW_PATCHELF_RB"])
```
Leave this as-was.

emctackett commented:
> Looking good so far! Can you suggest some test uses of `brew bump` you've been doing and I'll try it out locally? Thanks!

Hi Mike!  I've added a `--limit=` flag to `brew bump`, making it easier to test locally.

You can run something like `brew bump --limit=5` and expect to see final output like so:
```
==> Outdated formulae

==> aacgain
Current formula version: 1.8
Latest repology version: 1.9.0
Latest livecheck version: 1.8
Open pull requests: None.
==> acme
Current formula version: 0.96.5
Latest repology version: 2019.08.19
Latest livecheck version: Not found.
Open pull requests: None.
==> acmetool
Current formula version: 0.0.67
Latest repology version: 0.2.1
Latest livecheck version: 0.2.1
Open pull requests: {:title=>"acmetool: remove gopath and add caveat", :url=>"https://github.com/Homebrew/homebrew-core/pull/57766"}
==> acpica
Current formula version: 20200528
Latest repology version: 20200717
Latest livecheck version: 20200717
Open pull requests: None.
==> aescrypt
Current formula version: 0.7
Latest repology version: 3.14
Latest livecheck version: Not found.
Open pull requests: None.

```

Note that a hash of any open PRs detected for a package will be returned with the associated URL.

MikeMcQuaid commented:
@emctackett Keep up the good work! Will pull this locally and try it out when you let me know (either when CI is green or when you're stuck with the CI status [having rerun it once to get around anything flaky]). Thanks again!

emctackett commented:
Thanks, @MikeMcQuaid 👍 .  The GH Actions CI failure seems to have been a one-off; those tests are all passing now.

I added a flag for the user to query the version status of a particular package.  Here are some examples of what I would suggest running locally to test things out:
```
# --formula flag with valid formula name; example:
brew bump --formula=openclonk
==> Verifying outdated repology package as Homebrew formula
==> Outdated formula

==> openclonk
Current formula version: 7.0
Latest repology version: 8.1
Latest livecheck version: 8.1
Open pull requests: {:title=>"openclonk 8.1", :url=>"https://github.com/Homebrew/homebrew-core/pull/51504"}

# --formula flag with invalid name; example:
brew bump --formula=invalidName
==> Requested formula invalidname is not valid Homebrew formula.

# --formula flag with name of up-to-date formula:
brew bump --formula=agda
==> The requested formula, agda, is up-to-date.
Current version: 2.6.1

# --limit flag; example:
brew bump --limit=1
==> Querying outdated packages from Repology
==> 445 outdated packages identified
==> Verifying outdated repology packages as Homebrew formulae
==> Outdated formulae

==> aacgain
Current formula version: 1.8
Latest repology version: 1.9.0
Latest livecheck version: 1.8
Open pull requests: None.

# no flags
brew bump
# result will be equivalent to the above example, except that all outdated packages will be displayed.
# may run into GitHub API rate-limiting warnings if running locally.
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,138 @@
+    puts
+    outdated_packages.each do |formula, package_details|
+      ohai formula
+      puts "Current formula version: #{package_details[:current_formula_version]}"
+      puts "Latest repology version: #{package_details[:repology_latest_version]}"
```
```suggestion
      puts "Current formula version:  #{package_details[:current_formula_version]}"
      puts "Latest Repology version:  #{package_details[:repology_latest_version]}"
```

(to align better with `livecheck` version)

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,138 @@
+      ohai formula
+      puts "Current formula version: #{package_details[:current_formula_version]}"
+      puts "Latest repology version: #{package_details[:repology_latest_version]}"
+      puts "Latest livecheck version: #{package_details[:livecheck_latest_version] || "Not found."}"
```
```suggestion
      puts "Latest livecheck version: #{package_details[:livecheck_latest_version] || "Not found"}"
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,138 @@
+      puts "Current formula version: #{package_details[:current_formula_version]}"
+      puts "Latest repology version: #{package_details[:repology_latest_version]}"
+      puts "Latest livecheck version: #{package_details[:livecheck_latest_version] || "Not found."}"
+      puts "Open pull requests: #{package_details[:open_pull_requests] || "None."}"
```
```suggestion
      puts "Open pull requests: #{package_details[:open_pull_requests] || "None"}"
```

> Open pull requests: {:title=>"openclonk 8.1", :url=>"https://github.com/Homebrew/homebrew-core/pull/51504"}

Would also be nice to output this in e.g. `openclonk 8.1 (https://github.com/Homebrew/homebrew-core/pull/51504)` format (and make use of `Formatter.url`)

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,138 @@
+    requested_formula&.downcase!
+
+    if requested_formula && !get_formula_details(requested_formula)
+      ohai "Requested formula #{requested_formula} is not valid Homebrew formula."
```
```suggestion
      raise FormulaUnavailableError, requested_formula
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,138 @@
+    end
+
+    if requested_formula && outdated_repology_packages.nil?
+      ohai "The requested formula, #{requested_formula}, is up-to-date."
```
```suggestion
      ohai "#{requested_formula} is up-to-date!"
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,138 @@
+
+    if requested_formula && outdated_repology_packages.nil?
+      ohai "The requested formula, #{requested_formula}, is up-to-date."
+      puts "Current version: #{get_formula_details(requested_formula).version}"
```
I think it's still worth outputting the Repology and livecheck versions here (just to be explicit about whether they were checked/missing).

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,55 @@
+      page_no += 1
+    end
+
+    ohai "#{outdated_packages.size} outdated #{"package".pluralize(outdated_packages.size)} identified"
```
```suggestion
    puts "#{outdated_packages.size} outdated #{"package".pluralize(outdated_packages.size)} found"
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,138 @@
+  end
+
+  def validate_and_format_packages(outdated_repology_packages)
+    ohai "Verifying outdated repology #{"package".pluralize(outdated_repology_packages.size)} " \
```
```suggestion
```

Think this can be removed entirely because it's not time-consuming. Thoughts?

BaffourAdu replied:
I do agree with you @MikeMcQuaid but think it would be helpful when a verbose flag is passed. What do you think @emctackett

emctackett replied:
Hmm I don't have strong feelings about it but I'm leaning toward remove.  This line of output doesn't really provide the user with any useful information.  Since it's quick enough that the user isn't left wondering, it probably isn't necessary.

MikeMcQuaid replied:
Agreed @emctackett

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,138 @@
+  end
+
+  def display(outdated_packages)
+    ohai "Outdated #{"formula".pluralize(outdated_packages.size)}"
```
```suggestion
    ohai "Outdated #{"formula".pluralize(outdated_packages.size)}"
```

Think this can be removed as it's apparent from the output above.

MikeMcQuaid commented on Library/Homebrew/utils.rb:
```diff
@@ -10,8 +10,10 @@
+require "utils/repology"
 require "utils/svn"
 require "utils/tty"
+require "utils/repology"
```
```suggestion
```

BaffourAdu replied:
@MikeMcQuaid are you suggesting we remove this from `utils.rb` to `bump.rb`?

MikeMcQuaid replied:
No, sorry, just that it's duplicated 😅

BaffourAdu replied:
Okay, thanks for pointing that out.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,138 @@
+    display(outdated_packages)
+  end
+
+  def validate_and_format_packages(outdated_repology_packages)
```
Might be worth moving some of this method (and those below) into a module (either one of the existing ones or a new one) so they can be more easily tested.

emctackett commented:
@MikeMcQuaid thanks for the review!  Sounds great and I will work on all of those tasks this week :tada:.  I will update you when those are complete or if I run into trouble.

MikeMcQuaid commented:
@emctackett Perfect, thanks again!

emctackett commented:
@MikeMcQuaid just pinging you here so that you can take a look.  We've got some tests written and all checks are passing.

MikeMcQuaid commented on Library/Homebrew/test/utils/livecheck_formula_spec.rb:
```diff
@@ -0,0 +1,35 @@
+require "utils/livecheck_formula"
+
+describe LivecheckFormula do
+  describe "init", :integration_test do
```
Ideally this would not be an integration test. Are you sure that's required?

MikeMcQuaid commented on Library/Homebrew/test/utils/repology_spec.rb:
```diff
@@ -0,0 +1,73 @@
+require "utils/repology"
+
+describe Repology do
+  describe "formula_data", :integration_test do
```
Ideally this would not be an integration test. Are you sure that's required?

emctackett replied:
@MikeMcQuaid The reason for the `:integration_test` is because it seems to be required in order to use `install_test_formula "testball"`, which we are using as our test formula for some of these tests.  I tried to model this off of some of the other existing code seen in files like `formula_info_spec.rb`.  

Is there a better solution?  Thanks so much for your feedback!

BaffourAdu replied:
@MikeMcQuaid is this the best way to do it?
```ruby
    let(:f) { formula { url "foo-1.0" } }
    let(:options) { FormulaInstaller.new(f).display_options(f) }
    let(:action)  { "#{f.full_name} #{options}".strip }
```

MikeMcQuaid replied:
Yup, something like that would work. You can specify anything using `formula { }` .

BaffourAdu replied:
Great! Thanks for the feedback

MikeMcQuaid commented on Library/Homebrew/test/utils/repology_spec.rb:
```diff
@@ -0,0 +1,73 @@
+    end
+  end
+
+  describe "format_package", :integration_test do
```
Ideally this would not be an integration test. Are you sure that's required?

MikeMcQuaid commented on Library/Homebrew/test/dev-cmd/bump_spec.rb:
```diff
@@ -0,0 +1,7 @@
+
+describe "Homebrew.bump_args" do
+  it_behaves_like "parseable arguments"
+end
```
Might be nice to have a basic integration test in this file for `brew bump` .

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,70 @@
+  def display(formulae)
+    puts
+    formulae.each do |formula, package_details|
+      title = (up_to_date?(package_details) ? formula + " is up to date!" : formula).to_s
```
```suggestion
      title = (up_to_date?(package_details) ? "#{formula} is up to date!" : formula).to_s
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,70 @@
+    args = bump_args.parse
+
+    requested_formula = args.formula
+    requested_limit = args.limit ? args.limit.to_i : nil
```
```suggestion
    requested_limit = args.limit.to_i if args.limit.present?
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,70 @@
+
+    requested_formula = args.formula
+    requested_limit = args.limit ? args.limit.to_i : nil
+    requested_formula&.downcase!
```
In what situation would this not be downcased already? All formulae names are, I believe.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,69 @@
+  def bump
+    args = bump_args.parse
+
+    requested_formula = args.formula
```
```suggestion
    requested_formula = args.named
```

MikeMcQuaid replied:
Or ideally use `args.formulae.each` and loop between multiple?

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,69 @@
+        Display out-of-date brew formulae and the latest version available.
+        Also displays whether a pull request has been opened with the URL.
+      EOS
+      flag "--formula=",
+           description: "Return results for package by name."
```
```suggestion
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,69 @@
+    end
+
+    if repology_data.blank?
+      ohai "No Repology data found."
```
I think it'd be cool to still display the formula, livecheck and PR data in this case.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,69 @@
+  end
+
+  def display(formulae)
+    puts
```
```suggestion
```

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,111 @@
+    end
+
+    puts "#{outdated_packages.size} outdated #{"package".pluralize(outdated_packages.size)} found"
+
```
```suggestion
    puts

```

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,111 @@
+    pull_requests = GitHub.fetch_pull_requests(formula_name, tap_full_name, state: "open")
+
+    if pull_requests.try(:any?)
+      pull_requests = pull_requests.map { |pr| "#{pr["title"]} (#{Formatter.url(pr["url"])})" }.join(", ")
```
```suggestion
      pull_requests = pull_requests.map { |pr| "#{pr["title"]} (#{Formatter.url(pr["html_url"])})" }.join(", ")
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,69 @@
+  def bump_args
+    Homebrew::CLI::Parser.new do
+      usage_banner <<~EOS
+        `bump` [<options>]
```
```suggestion
        `bump` [<options>] [<formula>]
```

MikeMcQuaid commented:
Oops, think you may need a `brew update` to remove the manpage changes from the old Homebrew/homebrew-bundle with `brew man`.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,86 @@
+    formulae.each do |formula, package_details|
+      title = (up_to_date?(package_details) ? "#{formula} is up to date!" : formula).to_s
+      ohai title
+      puts "Current formula version:  #{package_details[:current_formula_version]}"
+      puts "Latest Repology version:  #{package_details[:repology_latest_version] || "Not found"}"
+      puts "Latest livecheck version: #{package_details[:livecheck_latest_version] || "Not found"}"
+      puts "Open pull requests: #{package_details[:open_pull_requests] || "None"}"
```
```suggestion
      puts "Current formula version:  #{package_details[:current_formula_version]}"
      puts "Latest Repology version:  #{package_details[:repology_latest_version] || "not found"}"
      puts "Latest livecheck version: #{package_details[:livecheck_latest_version] || "not found"}"
      puts "Open pull requests:       #{package_details[:open_pull_requests] || "none"}"
```

MikeMcQuaid commented on Library/Homebrew/utils/livecheck_formula.rb:
```diff
@@ -0,0 +1,27 @@
+  def init(formula)
+    ohai "Checking livecheck formula: #{formula}" if Homebrew.args.verbose?
+
+    response = Utils.popen_read(HOMEBREW_BREW_FILE, "livecheck", formula, "--quiet").chomp
```
CC @nandahkrishna That, once merged, this would be a good place to call into the Livecheck code rather than shell out.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump-formula-pr.rb:
```diff
@@ -298,6 +298,8 @@ def bump_formula_pr
 
     new_formula_version = formula_version(formula, requested_spec, new_contents)
 
+    check_duplicate_pull_requests(formula, tap_full_name, new_formula_version.to_s)
```
```suggestion
```

As this has now been integrated.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump-formula-pr.rb:
```diff
@@ -298,6 +298,7 @@ def bump_formula_pr
 
     new_formula_version = formula_version(formula, requested_spec, new_contents)
 
+
```
```suggestion
```

MikeMcQuaid commented on Library/Homebrew/utils/repology.rb:
```diff
@@ -0,0 +1,112 @@
+      pull_requests = pull_requests.map { |pr| "#{pr["title"]} (#{Formatter.url(pr["html_url"])})" }.join(", ")
+    end
+
+    pull_requests = "None" if pull_requests.empty?
```
```suggestion
    pull_requests = "none" if pull_requests.blank?
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,86 @@
+      puts "Current formula version:  #{package_details[:current_formula_version]}"
+      puts "Latest Repology version:  #{package_details[:repology_latest_version] || "not found"}"
+      puts "Latest livecheck version: #{package_details[:livecheck_latest_version] || "not found"}"
+      puts "Open pull requests:       #{package_details[:open_pull_requests] || "none"}"
```
When is this output vs. the `Open pull requests:` below? Would be nice to unify them if possible.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,86 @@
+  def bump
+    args = bump_args.parse
+
+    requested_formulae = !args.formulae.empty? ? args.formulae.map(&:name) : nil
```
```suggestion
    requested_formulae = args.formulae.map(&:name) if args.formulae.present?
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,86 @@
+
+      response
+    else
+      Repology.parse_api_response
```
This seems like it should take the `requested_limit` parameter because otherwise it can query e.g. three pages of results (which is a bit slow).

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,86 @@
+
+    repology_data = if requested_formulae
+      response = {}
+      requested_formulae.each do |formula|
```
If this is larger than `requested_limit` it feels like it should either error or break the loop.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -0,0 +1,86 @@
+
+    validated_formulae = {}
+
+    validated_formulae = Repology.validate_and_format_packages(repology_data, requested_limit) if repology_data
```
Might be nice to output something at this stage otherwise I can see:

```
==> Querying outdated packages from Repology
438 outdated packages found
```

and then an indefinite pause without seeing what is happening.

emctackett replied:
@MikeMcQuaid we previously had some `ohai` output at this stage and it was discussed here:

https://github.com/Homebrew/brew/pull/7834#discussion_r462230052

Ultimately it was removed after this discussion.  I can add it back if you think that's the better choice.

MikeMcQuaid replied:
Perhaps it's worth adding back only in the cases where it's going to be slow e.g. where you don't have a single formula?

MikeMcQuaid commented:
```
==> Querying outdated packages from Repology
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/2.4.12-125-g5d9e1f2\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.6\)\ curl/7.64.1 --retry 3 https://repology.org/api/v1/projects/\?inrepo=homebrew\&outdated=1
==> Paginating Repology API page: 1
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/2.4.12-125-g5d9e1f2\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.6\)\ curl/7.64.1 --retry 3 https://repology.org/api/v1/projects/libdsk/\?inrepo=homebrew\&outdated=1
==> Paginating Repology API page: 1
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/2.4.12-125-g5d9e1f2\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.6\)\ curl/7.64.1 --retry 3 https://repology.org/api/v1/projects/unibilium/\?inrepo=homebrew\&outdated=1
==> Paginating Repology API page: 1
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/2.4.12-125-g5d9e1f2\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.6\)\ curl/7.64.1 --retry 3 https://repology.org/api/v1/projects/yle-dl/\?inrepo=homebrew\&outdated=1
1 outdated package found

/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/aacgain.rb
==> Checking livecheck formula: aacgain
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/2.4.12-125-g5d9e1f2\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.6\)\ curl/7.64.1 --retry 3 --location https://api.github.com/search/issues\?q=aacgain\+repo\%3AHomebrew\%2Fhomebrew-core\+state\%3Aopen\+in\%3Atitle\&per_page=100 --header Accept:\ application/vnd.github.v3\+json --write-out '
'\%\{http_code\} --header Accept:\ application/vnd.github.antiope-preview\+json --user MikeMcQuaid:****** --dump-header /private/tmp/github_api_headers20200819-40296-1g5dw3p
==> aacgain
Current formula version:  1.8
Latest Repology version:  1.9.0
Latest livecheck version: 1.8
Open pull requests:       none
```

I'm maybe not understanding what `--limit=1` is doing here but I would have expected it to only get a single outdated from Repology and output that but it still seems to be working through multiple pages.

```
mikebook # brew bump --debug --verbose                 (mlh-outdated-packages) [~/Documents/homebrew]
==> Querying outdated packages from Repology
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/2.4.12-125-g5d9e1f2\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.6\)\ curl/7.64.1 --retry 3 https://repology.org/api/v1/projects/\?inrepo=homebrew\&outdated=1
==> Paginating Repology API page: 1
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/2.4.12-125-g5d9e1f2\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.6\)\ curl/7.64.1 --retry 3 https://repology.org/api/v1/projects/libdsk/\?inrepo=homebrew\&outdated=1
==> Paginating Repology API page: 1
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/2.4.12-125-g5d9e1f2\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.6\)\ curl/7.64.1 --retry 3 https://repology.org/api/v1/projects/unibilium/\?inrepo=homebrew\&outdated=1
==> Paginating Repology API page: 1
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/2.4.12-125-g5d9e1f2\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.6\)\ curl/7.64.1 --retry 3 https://repology.org/api/v1/projects/yle-dl/\?inrepo=homebrew\&outdated=1
428 outdated packages found

==> Verifying outdated repology packages
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/aacgain.rb
==> Checking livecheck formula: aacgain
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/2.4.12-125-g5d9e1f2\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.6\)\ curl/7.64.1 --retry 3 --location https://api.github.com/search/issues\?q=aacgain\+repo\%3AHomebrew\%2Fhomebrew-core\+state\%3Aopen\+in\%3Atitle\&per_page=100 --header Accept:\ application/vnd.github.v3\+json --write-out '
'\%\{http_code\} --header Accept:\ application/vnd.github.antiope-preview\+json --user MikeMcQuaid:****** --dump-header /private/tmp/github_api_headers20200819-40754-10m5sau
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/acme.rb
```

Similarly, in this case where it's going to iterate through thousands of formulae: how about outputting the `Current formula version` etc. for each formula as it works through so there's some useful output for each as it progresses?

To be clear: the `brew bump babel` etc. single formula output is now perfect 🎉

emctackett commented:
@MikeMcQuaid you are right, it looks like that was a bug that introduced.  I've addressed it now so that we don't paginate through more repology packages than necessary and the output is like so (I removed some excess debugging output for clarity):

```
brew bump --limit=1 --debug --verbose
==> Querying outdated packages from Repology
# ..
==> Paginating Repology API page: 1
# ..
399 outdated packages found

# ...
==> Checking livecheck formula: aacgain
# ...
==> aacgain
Current formula version:  1.8
Latest Repology version:  1.9.0
Latest livecheck version: not found
Open pull requests:       none
```

I will take a look at implementing the second part of your comment.

Also, you may have noticed but I was able to implement requesting multiple single formulae by name.  Example:

```
brew bump openclonk babel
==> openclonk
Current formula version:  7.0
Latest Repology version:  8.1
Latest livecheck version: not found
Open pull requests:       none
==> babel
Current formula version:  7.10.4
Latest Repology version:  not found
Latest livecheck version: not found
Open pull requests:       none
```

emctackett commented:
Ah I was also able to cleanup the way we query repology a bit so the example posted above now only grabs half of the amount of data from Repology.  Unfortunately, 200 is going to be the smallest amount of packages we can receive a response on if we aren't querying a package by name.

```
brew bump --limit=1 --debug --verbose
==> Querying outdated packages from Repology
# ..
==> Paginating Repology API page: 1
# ..
399 outdated packages found

# ...
==> Checking livecheck formula: aacgain
# ...
==> aacgain
Current formula version:  1.8
Latest Repology version:  1.9.0
Latest livecheck version: not found
Open pull requests:       none
```

MikeMcQuaid commented:
Thanks for your fantastic work on this PR @emctackett and @BaffourAdu and for all your patience here :clap:

emctackett commented:
Thanks, Mike!
